### ISSUE TITLE
🚀 Feature:  Security Scans like SAST, DAST, FOSS, CAST in pipeline

### ISSUE THREAD

**Issue by MananShah9** [appwrite/appwrite#5232]
### 🔖 Feature description

In the build process, all the security scans like SAST DAST FOSS and CAST should be performed to deliver a safer product

### 🎤 Pitch

While making appwrite safer and a product of a higher quality it opens up the possibility for many MNC's to adopt the product and thereby increasing the impact and reach of the same.

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by stnguyen90** [comment]
@MananShah9, thanks for creating this issue! 🙏🏼 Do you have suggestions on what to use? If you do, please also provide alternatives and the reason for choosing that tool over the alternative.

**Comment by MananShah9** [comment]
So I have used IBM App scan, checkmarx, Sonatype and netsparker before... I'm not really sure if I have reasons ready to choose one over other and might need to get involved a bit more to be able to brainstorm the same.

**Comment by joeyouss** [comment]
Hi @MananShah9 thanks for the suggestion. Do you mind clarifying for the community, including myself, what these scans do and why they'd be beneficial? Some of us without the background can't properly appreciate your suggestions, because we might not understand these terms.
This way, we can have more community participation in the discussion. Thanks!

**Comment by MananShah9** [comment]
These scans are like security checkups for softwares.

Think of them as tools that look closely at the source code or the environment to interact with the running app to find any weak points that could be exploited. 

Here's what each of these Scan does:

SAST (Static Application Security Testing): This tool reads the source code and figures out where there might be security problems.

DAST (Dynamic Application Security Testing): This tool watches the software while it's running and finds problems that only show up when the software is actually being used (only for web apps).

CAST (Container Scanning): This tool looks inside docker containers to find any issues that might be hiding there.

FOSS Scans (Free Open-Source Software Scans): This tool helps us find weaknesses in the free software packages we use in our own software.

It's really hard for us to find these problems just by looking at the code ourselves. 

The company I work for (HSBC Technologies) uses tools like the ones you can find here:

https://www.invicti.com/
https://checkmarx.com/
https://www.sonatype.com/products/vulnerability-scanner

I can't emphasize enough how important this scanning is. Pretty sure @stnguyen90 will back me here.
I have tried to keep this as simple as possible but do let me know if any further explanation is required

**Comment by stnguyen90** [comment]
We would like to have:

1. Secret Detection - Gitleaks <https://github.com/gitleaks/gitleaks>
   - good usability - multiple running options and config file
2. Dependency Check  - OSV by Google <https://osv.dev/>
   - good usability - JSON output and multiple options
   - good accuracy - OSV database from google and support for multiple languages including PHP
3. Infrastructure Misconfiguration - KICS <https://www.kics.io/>
   - good usability - multiple file types and powerful queries
   - good accuracy - nightly builds
4. Container Scanning - Trivy <https://trivy.dev/>
   - good usability - easy to configure
   - good accuracy - can detect multiple OS packages
5. Runtime Scanning - ZAP <https://owasp.org/www-project-zap/>
   - good usability - has a docker image and UI and is very extendable 
   - good accuracy - numersous features including sql injection fuzzing, broken access, XSS

**Comment by btme0011** [comment]
can i work on this feature?

**Comment by AvaterClasher** [comment]
Can i be assigned this ? Though i would like some mentoring in tackling this problem.

**Comment by btme0011** [comment]
> can i work on this feature?
Hey Haimantika can you assign me this issue. I am already working on it.

**Comment by tessamero** [comment]
hi @btme0011 , you've been assigned to the issue. You may now start working on it. :) Please keep us updated every 3 days so we know you are still working on it, so the issue doesn't get reassigned :)

happy coding 😃

**Comment by btme0011** [comment]
Will be raising seperate CR for each feature

**Comment by drraghavendra** [comment]
sir @tessamero can i work on this Pull Request as i will provide the source code

**Comment by brantheeman** [comment]
Can I work on the Dependency Check feature? @stnguyen90

**Comment by drraghavendra** [comment]
Security scans like SAST, DAST, FOSS, and CAST are used to identify and mitigate security vulnerabilities in software applications.

SAST (Static Application Security Testing)

SAST is a type of security testing that analyzes source code to identify security vulnerabilities. SAST tools can scan code for a variety of common vulnerabilities, such as SQL injection, cross-site scripting, and insecure coding practices.

DAST (Dynamic Application Security Testing)

DAST is a type of security testing that analyzes a running application to identify security vulnerabilities. DAST tools typically simulate real-world attacks to test the application's security.

FOSS (Free and Open Source Software)

FOSS is a type of software that is licensed under an open source license, which means that the source code is freely available for anyone to view, modify, and distribute. FOSS security scans are used to identify security vulnerabilities in FOSS software.

CAST (Compositional Analysis of Software Technology)

CAST is a type of security testing that analyzes the software supply chain to identify security vulnerabilities. CAST tools can scan software components, such as third-party libraries and frameworks, for security vulnerabilities.

When to use each type of security scan:

SAST: SAST should be used early in the software development lifecycle (SDLC) to identify and fix security vulnerabilities before they are introduced into the code.
DAST: DAST should be used later in the SDLC to test the security of the running application. DAST can be used to identify vulnerabilities that may not be detected by SAST, such as vulnerabilities that are caused by the interaction of multiple components of the application.
FOSS: FOSS security scans should be used to identify security vulnerabilities in FOSS software. FOSS security scans can be used to scan FOSS components before they are used in an application, or to scan existing applications for FOSS components with known vulnerabilities.
CAST: CAST should be used to identify security vulnerabilities in the software supply chain. CAST can be used to scan software components before they are used in an application, or to scan existing applications for components with known vulnerabilities.
Example use cases:

A software development team is developing a new web application. The team uses SAST to scan the source code for security vulnerabilities. The team also uses DAST to test the security of the running application.
A company is using a variety of FOSS components in its software applications. The company uses FOSS security scans to identify security vulnerabilities in the FOSS components.
A company is concerned about the security of its software supply chain. The company uses CAST to scan software components for security vulnerabilities before they are used in its applications.
It is important to note that no single type of security scan is sufficient to identify all security vulnerabilities in a software application. A comprehensive security testing program should include a mix of SAST, DAST, FOSS, CAST, and other security testing techniques.

**Comment by btme0011** [comment]
I am working on it :)

**Comment by btme0011** [comment]
1) Secret Detection - 
* Some of the work is require from the Admin side as i dont have the permission to add secrets.
      * Go to the Repo Settings
      * Go to secret and Variables
      * Go to Action
      * Add a Secret key with name GITLEAKS_LICENSE
      * How to get license -> [link](https://gitleaks.io/products.html)

**Comment by btme0011** [comment]
3) KICS installation - 
   got some initial issue while scanning with kics
   option 1 -> Remove these scans from the queries for now 
   option 2 -> Create a new issue and add these to that
   option 3 -> Dont implement Kicks
   option 4 -> Fix these issues also

issues -> https://github.com/btme0011/appwrite/actions/runs/6424134013/job/17444111351

**Comment by stnguyen90** [comment]
Please add the scan, but please create an issue to address the findings from the scan.

**Comment by tessamero** [comment]
@btme0011 the PR is still awaiting a follow up on your end before we can merge :)

**Comment by tessamero** [comment]
@btme0011 sounds good, I did send out a default message,  please check my message on the PR :)

**Comment by adityaoberai** [comment]
Closing this issue. Thank you so much for participating in Hacktoberfest 2023! We can't wait to welcome you all during HF 2024! Stay tuned for a lot more amazing issues from the Appwrite team!
